src/thread/thread_worker_pool.hpp

```diff
@@ -350,134 +631,692 @@ namespace cubthread
+
+    // either get a queued task or add to free active list
+    // note: returned task cannot be saved directly to m_task_p. if worker is added to wait queue and NULL is returned,
+    //       current thread may be preempted. worker is then claimed from free active list and worker is assigned
+    //       a task. this changes expected behavior and can have unwanted consequences.
+    task_type *task_p = m_parent_core->get_task_or_add_to_free_active_list (*this);
+    if (task_p != NULL)
+      {
+	m_statistics.collect_and_time (wpstat::id::SEARCH_TASK_IN_QUEUE, m_time_point);
+
+	// it is safe to set here
+	m_task_p = task_p;
+	return true;
+      }
+
+    // wait for task
+    std::unique_lock<std::mutex> ulock (m_task_mutex);
+    if (m_task_p == NULL && !m_stop)
+      {
+	// wait until a task is received or stopped ...
+	// ... or time out
+	m_task_cv.wait_for (ulock, WAIT_TIME, [this] { return m_task_p != NULL || m_stop; });
+      }
+    else
       {
-	// no threads available
-	THREAD_WP_LOG ("register_worker", "thread_p = %p", NULL);
-	return NULL;
+	// no need to wait
       }
+    // unlock mutex
+    ulock.unlock ();
 
-    // if thread was already used, we must join it before reusing
-    if (thread_p->joinable ())
+    if (m_task_p == NULL)
       {
-	thread_p->join ();
+	// not task received, remove from active list
+	if (m_parent_core->remove_worker_from_active_list (*this))
+	  {
+	    // removed from active list
+	  }
+	else
+	  {
+	    // I was not removed from active list... that means somebody claimed me right before removing from list
+	    // now I have to wait for the task
+	    ulock.lock ();
+	    if (m_task_p == NULL)
+	      {
+		// wait until task is set
+		m_task_cv.wait (ulock, [this] { return m_task_p != NULL; });
+	      }
+	    else
+	      {
+		// task already set, don't wait
+	      }
+	    ulock.unlock ();
+	  }
       }
 
-    ++m_worker_count;
+    if (m_task_p == NULL)
+      {
+	// no task assigned
+	m_time_point = wpstat::clock_type::now ();
 
-    THREAD_WP_LOG ("register_worker", "thread = %zu", get_thread_index (*thread_p));
-    return thread_p;
+	return false;
+      }
+    else
+      {
+	// found task
+	m_time_point = m_push_time;
+	m_statistics.collect_and_time (wpstat::id::WAKEUP_WITH_TASK, m_time_point);
+
+	return true;
+      }
   }
 
   template <typename Context>
-  inline void
-  worker_pool<Context>::deregister_worker (std::thread &thread_arg)
+  void
+  worker_pool<Context>::core::worker::run (void)
   {
-    // THREAD_WP_LOG ("deregister_worker", "thread = %zu", get_thread_index (thread_arg));
-    // no logging here; no thread & error context
-    m_thread_dispatcher.retire (thread_arg);
-    --m_worker_count;
+    task_type *task_p = NULL;
+
+    while (true)
+      {
+	init_run ();    // do stuff at the beginning like creating context
+
+	// loop and execute as many tasks as possible
+	do
+	  {
+	    execute_current_task ();
```

Comment:
We need to `recycle context` here to execute another task.

Reply:
You are right. I pushed a fix.